vehicle/OVMS.V3/components/vehicle_fiatedoblo/src/vehicle_fiatedoblo.cpp

```diff
@@ -0,0 +1,557 @@
+}
+
+// TODO add DobloXMsg_0x5c3_cap?
+// DobloXMsg_0x563
+uint8_t last_data[8];
+uint8_t dd[8];
+uint8_t dd2[8];
+
+// andle incoming data on StdOBD CAN
+//412 is sent also while charging (start)
+void OvmsVehicleFiatEDoblo::IncomingFrameCan1(CAN_frame_t* p_frame)
+{
+  uint8_t *d = p_frame->data.u8;
+
+  lastCanFrameTime = monotonictime;
+  if(IsPollState_Off()){
+    ESP_LOGI(TAG, "waking up by CAN frame %x", p_frame->MsgID);
+    PollState_Charging();
+  }
+  
+  switch (p_frame->MsgID){
+  case 0x305: // steering wheel angle in byte 0 and 1 (x10) maybe not, as it's in 2eb
+  case 0x30d: // wheel rotation speed
+    {
+      break;
+    }
+  case 0x38d: // vehicle speed (bytes 0 and 1 x100kmh)
+    {
+       int changed = 0;
+       uint8_t mask[8] = {0,0,0,0,0,0xff,0,0};
+       float speed;
+       for (int i = 0; i < 8; i++){
+	 if((dd2[i] & ~mask[i]) != (d[i] & ~mask[i])){
+	  dd2[i] = d[i];
+	  changed = 1;
+	 }
+       }
+       
+       if(changed){
+	 // TODO: check
+	 speed = ((((uint32_t)d[0] << 8) + d[1]) / 100.0);
+	 ESP_LOGI(TAG, "changes in 0x%4x: 0x%2x 0x%2x 0x%2x 0x%2x  0x%2x 0x%2x 0x%2x 0x%2x   speed = %f km/h", p_frame->MsgID, d[0], d[1], d[2], d[3], d[4], d[5], d[6], d[7], speed); // TODO remove
+       }
+       break;
+    }
+  case 0x3a8:
+    {
+      // TODO       ESP_LOGI(TAG, "received SoC: %f", (float)d[3] + ((float)d[4] / 256.0));
+
+      StandardMetrics.ms_v_bat_soc->SetValue((float)d[3] + ((float)d[4] / 256.0));
+    }
+    break;
+  case 0x552:
+    {
+      // TODO: which odometer setting is correct
+      StandardMetrics.ms_v_pos_odometer->SetValue((float)(((uint32_t)d[4] << 16) + ((uint32_t)d[5] << 8) + d[6]));
+    }
+    break;
+   case 0x2eb:
+     {
+       // steering wheel angle in bytes 0 and 1: negative: right, positive: left
+       // angle = ((uint32_t)d[0] << 8) + d[1]) / 10.0);
+     }
+     break;
+   case 0x3ad:
+     {
+       int changed = 0;
+       StandardMetrics.ms_v_env_awake->SetValue(true);
+
+       /*
+       if(d[6] & 0x40) Schalter Feststellbremse anziehen betätigt
+       if(d[6] & 0x10) Schalter Feststellbremse lösen betätigt
+
+       if(d[4] & 0x01) Feststellbremse angezogen
+       if(d[4] & 0x02) Feststellbremse Motor läuft (Zustand wird sich gleich ändern)
+       d[7] is CNT/CRC
+       */		   
+       uint8_t mask[8] = {0,0,0,0,0x03,0xf0,0x50,0xff}; // no idea what d[5].0xf0 is but the bits toggle often
+       for (int i = 0; i < 8; i++){
+	 if((dd[i] & ~mask[i]) != (d[i] & ~mask[i])){
+	  dd[i] = d[i];
+	  changed = 1;
+	 }
+       }
+       if(changed){
+	 ESP_LOGI(TAG, "changes in 0x%4x: 0x%2x 0x%2x 0x%2x 0x%2x  0x%2x 0x%2x 0x%2x 0x%2x", p_frame->MsgID, d[0], d[1], d[2], d[3], d[4], d[5], d[6], d[7]); // TODO remove
+       }
+       StandardMetrics.ms_v_env_handbrake->SetValue((d[4] & 0x01));
+       if((d[4] & 0x01) == 0){
+	 // handbrake released
+	 StandardMetrics.ms_v_env_parktime->SetValue(0);
```

Comment:
`ms_v_env_parktime` and `ms_v_env_drivetime` are derived from `ms_v_env_on` by the vehicle framework, no need to take care of these yourself. Same applies to the charge time. Generally always check first for automatisms & hooks provided by the vehicle base class before adding your own logic, to avoid introducing conflicts.

Reply:
I tried to spot the ignition signal already but failed to find it, also the charging state signal would be interesting. Can you point me to the docs of the RE toolkit? So far I mainly tried with SaavyCAN.